src/rust/receiver-mock/src/metrics.rs

```diff
@@ -21,71 +42,212 @@ pub fn handle_carbon2(
-                (*stats).metrics += 1;
-
+            let mut split = metric.split("=");
+            let field_name = split.nth(0).unwrap().to_string();
+            if field_name == "metric" {
+                // nth() consumes elements hence nth(0) again
+                let metric_name = split.nth(0).unwrap().to_string();
+                result.handle_metric(metric_name);
                 break;
             }
         }
 
-        let metrics_ip_list = (*stats).metrics_ip_list.entry(address).or_insert(0);
-        *metrics_ip_list += 1;
+        result.handle_ip(address);
     }
+
+    result
 }
 
 // Handle metrics in Graphite format
 // Reference: https://help.sumologic.com/Metrics/Introduction-to-Metrics/Metric-Formats#graphite
 pub fn handle_graphite(
     lines: std::str::Lines,
     address: IpAddr,
-    stats: &Arc<Mutex<Statistics>>,
     print_opts: options::Print,
-) {
-    let mut stats = stats.lock().unwrap();
+) -> MetricsHandleResult {
+    let mut result = MetricsHandleResult {
+        metrics: 0,
+        metrics_list: HashMap::new(),
+        metrics_ip_list: HashMap::new(),
+    };
 
     for line in lines {
         if print_opts.metrics {
             println!("metric => {}", line);
         }
-        let split = line.split(' ').collect::<Vec<_>>();
-        if split.len() != 3 {
+        let split_line = line.split(' ').collect::<Vec<_>>();
+        if split_line.len() != 3 {
             println!("Incorrect graphite metric line: {}", line);
             continue;
         }
 
-        let metric_name = split[0].split('.').last().unwrap().to_string();
-        let saved_metric = (*stats).metrics_list.entry(metric_name).or_insert(0);
-        *saved_metric += 1;
-        (*stats).metrics += 1;
+        let split_metric = split_line[0].split('.').collect::<Vec<_>>();
+        if split_metric.len() != 3 {
+            println!("Incorrect graphite metric name: {}", split_line[0]);
+            continue;
+        }
 
-        let metrics_ip_list = (*stats).metrics_ip_list.entry(address).or_insert(0);
-        *metrics_ip_list += 1;
+        // TODO: do we want to substitute dots with '_'s for graphite?
+        let metric_name = split_metric[1];
+        let metric_field = split_metric[2];
+        result.handle_metric(format!("{}_{}", metric_name, metric_field));
+        result.handle_ip(address);
     }
+
+    result
 }
 
 // Handle metrics in Prometheus format
 // Reference: https://help.sumologic.com/Metrics/Introduction-to-Metrics/Metric-Formats#prometheus
 pub fn handle_prometheus(
     lines: std::str::Lines,
     address: IpAddr,
-    stats: &Arc<Mutex<Statistics>>,
     print_opts: options::Print,
-) {
-    let mut stats = stats.lock().unwrap();
+) -> MetricsHandleResult {
+    let mut result = MetricsHandleResult {
+        metrics: 0,
+        metrics_list: HashMap::new(),
+        metrics_ip_list: HashMap::new(),
+    };
 
     for line in lines {
+        // Ignore comments
```

Comment:
Prometheus doesn't send comments. What does it came from?

Reply:
Prometheus format. I tested it with telegraf and I've seen comment lines sent

Reply:
so the telegraf unnecessary sends the comments. Could we address that in the telegraf repository? It affects performance/bandwidth

Reply:
Those comments have any meaning? Their structure seems to show that they might mean something

Reply:
it could, it specifies type 🤔